proposals/index-range-changes.md

```diff
@@ -0,0 +1,297 @@
+update existing collections using extension methods. This means the adoption of the feature will be slowed by waiting
+for library authors to update their types.
+
+The use of `Index` does come with a small performance hit. Typically a consumer of `Index` will pay an extra branch
+check (`IsFromEnd` and bounds) as they would with a `int` parameter (bounds only). Moving to `Range` doubles this cost
+as it occurs on each `Index`.
+
+This penalty is small and insignificant in the vast majority of code. For highly optimized code though this extra 
+branching can be unacceptable. It simply can't be thought of as a syntactic transformation from a call to 
+`span.Slice(start, end)` to `span[start..end]`. Instead the performance implications must be evaluated on each
+usage. This will likely lead to the feature being banned in certain portions of code bases.
+
+This proposals seeks to remove the abstraction penalties around `Index` and `Range`: both at an API and performance 
+layer. The goal being that most collection types just work today with minimal change and the feature can be adopted
+in existing code without fear of lost performance.
+
+This proposal specifically does not want to change how `Index` and `Range` type binding occurs. The index and range
+expressions continue to have the same syntax and types.
+
+## Detailed Design 
+### Countable Types
+Any type which has an property named `Length` or `Count` with an accessible getter adn a return type of `int` is 
+considered Countable. The language can make use of this property to convert an expression of type `Index` into an `int`
+at the point of the expression without the need to use the type `Index` at all. In the case both `Length` and `Count`
+are present then `Length` will be preferred. 
+
+For simplicity going forward the proposal will use the name `Length` to represent `Count` or `Length`.
+
+### Index and Range implementations are known
+The implementations of `Index` and `Range` are considered to be known and side effect free. Much like 
+`ValueTuple<T1, T2>` the language can assume a standard implementation and emit code inline which represents the 
+implementation of methods on `Index` and `Range`. This implementation includes methods like `GetOffset` or conversions
+like the implicit conversion from `int` to `Index`.
+
+All arithemtic operations which are emitted will be done so using an `unchecked` context. That matches the context
+in which `Index` and `Range` are compiled in.
+
+### Implicit Index support
+The language will provide an instance indexer member with a single parameter of type `Index` for types which meet the 
+following criteria:
+
+- The type is Countable
+- The type has an accessible instance indexer which takes a single `int` as the argument.
+- The type does not have an accessible instance indexer which takes a single `Index` as the parameter.
```

Comment:
any concerns about params in an indexer?

Reply:
You mean in the original indexer?

Reply:
yes.  say i had an indexer like so:  

```c#
int this[int i] { get; }
int this[Index x, int y = 0, int z = 0] { get; }
```
 
based on the above, it sounds like you'd generate `int this[Index] { get; }`  Which would then supercede my 'index-indexer'.  Maybe something we just don't care about.  But it's somehting i always notice when we say "takes N parameters".  There's always that bit of ambiguity if it means exactly that, or if it means "not counting optional/params parameters".  

Thanks!

Reply:
That's a good point. The intent of this is to provide an `Index` indexer if one wasn't available. This would certainly be applicable for the scenario hence probably shouldn't override it.